rectools/models/implicit_als.py

```diff
@@ -17,23 +17,104 @@
 
 import implicit.gpu
 import numpy as np
+import typing_extensions as tpe
+from implicit.als import AlternatingLeastSquares
 from implicit.cpu.als import AlternatingLeastSquares as CPUAlternatingLeastSquares
 from implicit.gpu.als import AlternatingLeastSquares as GPUAlternatingLeastSquares
 from implicit.utils import check_random_state
+from pydantic import BeforeValidator, ConfigDict, PlainSerializer, SerializationInfo, WrapSerializer
 from scipy import sparse
 from tqdm.auto import tqdm
 
 from rectools.dataset import Dataset, Features
 from rectools.exceptions import NotFittedError
+from rectools.models.base import ModelConfig
+from rectools.utils.config import BaseConfig
+from rectools.utils.misc import get_class_or_function_full_path, import_object
 
 from .rank import Distance
 from .vector import Factors, VectorModel
 
-AVAILABLE_RECOMMEND_METHODS = ("loop",)
 AnyAlternatingLeastSquares = tp.Union[CPUAlternatingLeastSquares, GPUAlternatingLeastSquares]
 
 
-class ImplicitALSWrapperModel(VectorModel):
+def _get_alternating_least_squares_class(spec: tp.Any) -> tp.Any:
+    if not isinstance(spec, str):  # including None
+        return spec
+    return import_object(spec)
+
+
+def _serialize_alternating_least_squares_class(
+    cls: tp.Optional[tp.Type[AnyAlternatingLeastSquares]], handler: tp.Callable, info: SerializationInfo
+) -> tp.Union[None, str, AnyAlternatingLeastSquares]:
+    if cls in (CPUAlternatingLeastSquares, GPUAlternatingLeastSquares) or cls is None:
+        return None
+    if info.mode == "json":
+        return get_class_or_function_full_path(cls)
+    return cls
+
+
+AlternatingLeastSquaresClass = tpe.Annotated[
+    tp.Optional[tp.Type[AnyAlternatingLeastSquares]],
+    BeforeValidator(_get_alternating_least_squares_class),
+    WrapSerializer(
+        func=_serialize_alternating_least_squares_class,
+        when_used="always",
+    ),
+]
+
+DType = tpe.Annotated[
+    np.dtype, BeforeValidator(func=np.dtype), PlainSerializer(func=lambda dtp: dtp.name, when_used="json")
+]
+
+
+def _serialize_random_state(rs: tp.Optional[tp.Union[None, int, np.random.RandomState]]) -> tp.Union[None, int]:
+    if rs is None or isinstance(rs, int):
+        return rs
+
+    # NOBUG: We can add serialization using get/set_state, but it's not human readable
+    raise TypeError("`random_state` must be ``None`` or have ``int`` type to convert it to simple type")
```

Comment:
If we return `None` or even a string like `RandomState object at ...`, then we will break our approach `Model.from_config(model.get_config()) == model`.
We can try to add additional param like `errors="ignore"`, but I don't know how to implement it (yet)

Reply:
Yes, we will break this logic (with a warning that is was broken).
And we need to decide the desired behaviour now because we will have this problem in LightFM soon.
And in save/load for LightFM too. Which is super important. What is the desired behaviour in LightFM?
And let's say we are not going with `Model.from_config(model.get_config())` but we are actually saving LightFM config somewhere (mlflow, locally) and then uploading it to create s model from it. This will be the most common case for config usage.

Reply:
save/load is a totally differen story - we can easily save random state in binary file

Actually we can save it to config as well, but it will contain array of 624 integers

Reply:
If we can avoid an error and it will support our approach `Model.from_config(model.get_config()) == model` - let's do it?

Reply:
Decided keep in this PR as is, and improve later with adding parameter to `get_config` method like `non_serializable: tp.Literal["error", "stub"] = "error"`